spec/rspec/mocks/matchers/receive_spec.rb

```diff
@@ -109,6 +123,18 @@ module Mocks
 
               expect(receiver.foo(kw: :arg)).to eq(:arg)
             end
+
+            it "expects to receive keyword args" do
+              dbl = instance_double(TestObject)
+              expect(dbl).to receive(:kw_args_method).with(a: 1, b: 2)
+              dbl.kw_args_method(a: 1, b: 2)
+            end
+
+            it "expects to receive keyword args with a hash" do
+              dbl = instance_double(TestObject)
+              expect(dbl).to receive(:kw_args_method).with({a: 1, b: 2})
+              dbl.kw_args_method(a: 1, b: 2)
```

Comment:
This should fail, shouldn't it? We expect an options hash, and pass kwargs.

Reply:
Good point. I am wondering if it is a side effect of changes in 3.2. Like here. @eregon any idea?

<img width="613" alt="Screen Shot 2023-01-03 at 10 09 41" src="https://user-images.githubusercontent.com/8417720/210327926-e6887a68-5946-4f43-9206-da49de5198e8.png">

Reply:
We seem to be in the opposite situation when an option hash passed as a positional argument to an expectation suddenly matches with a kwargs hash.

I suppose that the problem is here:
```ruby
            if Hash === expected_args.last && Hash === actual_args.last
              if !Hash.ruby2_keywords_hash?(actual_args.last) && Hash.ruby2_keywords_hash?(expected_args.last)
                return false
```
In our scenario, `Hash.ruby2_keywords_hash?(actual_args.last)` is `true`, while the other one is `false`.
They don't match, but we don't return `false` like we do e.g. here:
```ruby
          if Hash === expected_hash && Hash === actual_hash &&
            (Hash.ruby2_keywords_hash?(expected_hash) != Hash.ruby2_keywords_hash?(actual_hash))
```

Reply:
I suggest changing this:
```diff
-              if !Hash.ruby2_keywords_hash?(actual_args.last) && Hash.ruby2_keywords_hash?(expected_args.last)
+              if Hash.ruby2_keywords_hash?(actual_args.last) != Hash.ruby2_keywords_hash?(expected_args.last)
```

This is causing two more specs to fail, at least on Ruby 3.2:
```ruby
            it "matches against a hash submitted as keyword arguments and received as positional argument (in both Ruby 2 and Ruby 3)" do
              expect(dbl).to receive(:kw_args_method).with(1, {:required_arg => 2, :optional_arg => 3})
              dbl.kw_args_method(1, :required_arg => 2, :optional_arg => 3)
            end
```
```ruby
        it "matches against a hash submitted as keyword arguments a and received as a positional argument (in both Ruby 2 and Ruby 3)" do
          opts = {:a => "a", :b => "b"}
          expect(a_double).to receive(:random_call).with(opts)
          a_double.random_call(:a => "a", :b => "b")
        end
```

But if we'd make this comparison strict, I agree that it should fail.

Do you guys agree to make such a change, and see if it affects other Rubies?

Reply:
Now this is where I got stuck:
```ruby
               opts = {:a => "a", :b => "b"}
               expect(a_double).to receive(:random_call).with(opts)
               a_double.random_call(:a => "a", :b => "b")
```
fails with
```
     Failure/Error: a_double.random_call(:a => "a", :b => "b")
       #<Double (anonymous)> received unexpected message :random_call with ({:a=>"a", :b=>"b"})
```
but when I want to expect that failure:
```ruby
             expect do
               opts = {:a => "a", :b => "b"}
               expect(a_double).to receive(:random_call).with(opts)
               a_double.random_call(:a => "a", :b => "b")
               # RSpec::Mocks.space.verify_all
             end.to fail_with(/received unexpected message :random_call with \({:a=>"a", :b=>"b"}\)/)
```
it:
```
     Failure/Error: expect(a_double).to receive(:random_call).with(opts)

       (Double (anonymous)).random_call({:a=>"a", :b=>"b"})
           expected: 1 time with arguments: ({:a=>"a", :b=>"b"})
           received: 0 times
```
and adding `verify_all`/`RSpec::Mocks.verify` doesn't help.

What's even more peculiar is that if I expect a failure message that doesn't match:
```ruby
             expect do
               opts = {:a => "a", :b => "b"}
               expect(a_double).to receive(:random_call).with(opts)
               a_double.random_call(:a => "a", :b => "b")
               RSpec::Mocks.verify
             end.to fail_with(/foo/)
```
it seems to make sense:
```ruby
     Failure/Error:
       expect do
         opts = {:a => "a", :b => "b"}
         expect(a_double).to receive(:random_call).with(opts)
         a_double.random_call(:a => "a", :b => "b")
       end.to fail_with(/foo/)

       expected RSpec::Mocks::MockExpectationError with message matching /foo/, got #<RSpec::Mocks::MockExpectationError: #<Double (anonymous)> received unexpected message :random_call with ({:a=>"a", :b=>"b"})> with backtrace:
```
@JonRowe I ended up with:
```ruby
             expect do
               opts = {:a => "a", :b => "b"}
               expect(a_double).to receive(:random_call).with(opts)
               a_double.random_call(:a => "a", :b => "b")
             end.to fail_with do |failure|
               expect(failure.message).to match(/received unexpected message :random_call with \(\{:a=>"a", :b=>"b"\}\)/)
               reset_all
             end
```
and double-checked that this example fails if I pass a different failure message to `match`. Do you have a better suggestion?

Reply:
The problem is the comment above:
> # if both arguments end with Hashes, and if one is a keyword hash and the other is not, they don't match

is super confusing. That should be updated to something like:

```ruby
# if both arguments end with Hashes, and if #with was called with keywords but the method was called without keywords, they don't match
# if #with is called without keywords, e.g., with({a: 1}), then it's fine to call it with either foo(a: 1) or foo({a: 1}) because it's as if the method was: def foo(options).
```

Reply:
But this is only true for methods that accept positional parameters, right? E.g. for:
```ruby
class A
  def foo(bar:)
  end
end

a = A.new
expect(a).to receive(:foo).with({bar: 1})
a.foo({bar: 1})
```

If we follow
> if `#with` is called without keywords, e.g., `with({a: 1})`, then it's fine to call it with either `foo(a: 1)` or `foo({a: 1})` because it's as if the method was: `def foo(options)`.

The spec will pass, but the real-life code:
```ruby
a = A.new
a.foo({bar: 1})
```
would blow up with:
```
ArgumentError: wrong number of arguments (given 1, expected 0; required keyword: bar)
```

that means we would be concealing the error.

Is my understanding correct, @eregon ?

Reply:
It's a mock, so the method might not exist, right?
From `with({a: 1})` that doesn't indicate the method takes keyword arguments, hence it can be called with whatever (passing `a: 1` or `{a: 1}` shouldn't make any difference).

If the method does exist like in that case, I think maybe RSpec could warn that you call `with` without kwargs, but the method accepts kwargs.
Or, you could actually check whether the existing method accepts kwargs or not for the kwargs-vs-positional check, but then what happens if the method does not exist?

Wouldn't this specific example fail anyway because it actually calls the original method?
And if it doesn't call the original method, isn't it wrong to look at the original method?

Reply:
Also this example wouldn't be caught by your change, right? (no ruby2_keyword flag on both sides)
So it's a bug of writing `.with({bar: 1})` which means "expect no kwargs" when the real method only accepts kwargs. Hence maybe that's worth a warning.

Reply:
>  this example wouldn't be caught by your change

Fair enough. The rest makes total sense, too.